File: controllers/blog-crud-controller.js
const { validationResult } = require("express-validator");
const HttpError = require("../models/http-error");
const User = require("../models/user");
const Blog = require("../models/blog");
const mongoose = require("mongoose");
const createBlogPage = async (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return next(
      new HttpError("Invalid inputs passed, please check your data.", 422)
    );
  }
  const { title, description, article } = req.body;
  const userId = req.userData.userId;
  const createdBlog = new Blog({
    title,
    description,
    article,
    author: userId,
  });
  try {
    const session = await mongoose.startSession();
    session.startTransaction();
    await createdBlog.save({ session });
    const user = await User.findById(userId);
    user.blogs.push(createdBlog);
    await user.save({ session });
    session.commitTransaction();
  } catch (err) {
    const error = new HttpError(
      "Creating new blog failed, please try again later.",
      500
    );
    return next(error);
  }
  res.send({ createdBlog });
};
const updateBlogPage = async (req, res, next) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return next(
      new HttpError("Invalid inputs passed, please check your data.", 422)
    );
  }
  const { title, description, article } = req.body;
  const blogId = req.params.bid;
  try {
    const blog = await Blog.findById(blogId);
    if (blog.author != req.userData.userId) {
      return next(new HttpError("Unauthorized!", 401));
    }
    blog.title = title;
    blog.description = description;
    blog.article = article;
    await blog.save();
    res.send({ blog });
  } catch (err) {
    return next(
      new HttpError("Updating  blog failed, please try again later.", 500)
    );
  }
};
const deleteBlogPage = async (req, res, next) => {
  const blogId = req.params.bid;
  try {
    const session = await mongoose.startSession();
    const blog = await Blog.findById(blogId);
    if (blog.author != req.userData.userId) {
      return next(new HttpError("Unauthorized!", 401));
    }

    session.startTransaction();
    await Blog.findByIdAndDelete(blogId).setOptions({ session });
    await User.updateOne(
      { blogs: blogId },
      { $pull: { blogs: blogId } }
    ).setOptions({ session });
    await session.commitTransaction();

    res.send({ blog });
  } catch (err) {
    console.log(err);
    return next(
      new HttpError("Deleting blog failed, please try again later.", 500)
    );
  }
};
const getUsersPages = async (req, res, next) => {
  const id = req.userData.userId;
  try {
    const user = await User.findById(id).populate("blogs");
    res.send({
      blogs: user.blogs,
      author: { firstName: user.firstName, lastName: user.lastName },
    });
  } catch (error) {
    return next(new HttpError("failed to get user's blogs", 500));
  }
};
exports.createBlogPage = createBlogPage;
exports.updateBlogPage = updateBlogPage;
exports.deleteBlogPage = deleteBlogPage;
exports.getUsersPages = getUsersPages;

File: go_rewrite/controllers/blog-crud-controller.go
package controllers

import (
	"context"
	"time"

	"backend/config"
	"backend/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

func CreateBlog(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var blogReq models.BlogRequest
	if err := c.ShouldBindJSON(&blogReq); err != nil {
		c.JSON(422, gin.H{"message": "Invalid inputs passed, please check your data."})
		return
	}

	userData := c.MustGet("userData").(map[string]string)
	uid, _ := primitive.ObjectIDFromHex(userData["userId"])

	blog := models.Blog{
		Title:       blogReq.Title,
		Description: blogReq.Description,
		Article:     blogReq.Article,
		Author:      uid,
		Comments:    []primitive.ObjectID{},
	}

	result, err := config.BlogCollection.InsertOne(ctx, blog)
	if err != nil {
		c.JSON(500, gin.H{"message": "Creating new blog failed, please try again later."})
		return
	}

	_, err = config.UserCollection.UpdateOne(ctx, bson.M{"_id": uid}, bson.M{"$push": bson.M{"blogs": result.InsertedID}})
	if err != nil {
		c.JSON(500, gin.H{"message": "Creating new blog failed, please try again later."})
		return
	}

	c.JSON(201, gin.H{"createdBlog": blog})
}

func UpdateBlog(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	blogId := c.Param("bid")
	bid, err := primitive.ObjectIDFromHex(blogId)
	if err != nil {
		c.JSON(500, gin.H{"message": "Invalid blog ID"})
		return
	}

	var blogReq models.BlogRequest
	if err := c.ShouldBindJSON(&blogReq); err != nil {
		c.JSON(422, gin.H{"message": "Invalid inputs passed, please check your data."})
		return
	}

	userData := c.MustGet("userData").(map[string]string)
	uid, _ := primitive.ObjectIDFromHex(userData["userId"])

	var blog models.Blog
	err = config.BlogCollection.FindOne(ctx, bson.M{"_id": bid}).Decode(&blog)
	if err != nil {
		c.JSON(500, gin.H{"message": "Updating blog failed, please try again later."})
		return
	}

	if blog.Author != uid {
		c.JSON(401, gin.H{"message": "Unauthorized!"})
		return
	}

	_, err = config.BlogCollection.UpdateOne(ctx, bson.M{"_id": bid}, bson.M{"$set": bson.M{
		"title":       blogReq.Title,
		"description": blogReq.Description,
		"article":     blogReq.Article,
	}})
	if err != nil {
		c.JSON(500, gin.H{"message": "Updating blog failed, please try again later."})
		return
	}

	c.JSON(200, gin.H{"message": "Blog updated!"})
}

func DeleteBlog(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	blogId := c.Param("bid")
	bid, err := primitive.ObjectIDFromHex(blogId)
	if err != nil {
		c.JSON(500, gin.H{"message": "Invalid blog ID"})
		return
	}

	userData := c.MustGet("userData").(map[string]string)
	uid, _ := primitive.ObjectIDFromHex(userData["userId"])

	var blog models.Blog
	err = config.BlogCollection.FindOne(ctx, bson.M{"_id": bid}).Decode(&blog)
	if err != nil {
		c.JSON(500, gin.H{"message": "Deleting blog failed, please try again later."})
		return
	}

	if blog.Author != uid {
		c.JSON(401, gin.H{"message": "Unauthorized!"})
		return
	}

	_, err = config.BlogCollection.DeleteOne(ctx, bson.M{"_id": bid})
	if err != nil {
		c.JSON(500, gin.H{"message": "Deleting blog failed, please try again later."})
		return
	}

	_, err = config.UserCollection.UpdateOne(ctx, bson.M{"_id": uid}, bson.M{"$pull": bson.M{"blogs": bid}})
	if err != nil {
		c.JSON(500, gin.H{"message": "Deleting blog failed, please try again later."})
		return
	}

	c.JSON(200, gin.H{"message": "Blog deleted!"})
}

func GetUserBlogs(c *gin.Context) {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	userData := c.MustGet("userData").(map[string]string)
	uid, _ := primitive.ObjectIDFromHex(userData["userId"])
	firstName := userData["firstName"]
	lastName := userData["lastName"]

	cursor, err := config.BlogCollection.Find(ctx, bson.M{"author": uid})
	if err != nil {
		c.JSON(500, gin.H{"message": "failed to get user's blogs"})
		return
	}
	defer cursor.Close(ctx)

	var blogs []models.Blog = make([]models.Blog, 0)
	if err := cursor.All(ctx, &blogs); err != nil {
		c.JSON(500, gin.H{"message": "failed to get user's blogs"})
		return
	}

	c.JSON(200, models.UserBlogResponse{
		Blogs: blogs,
		Author: models.Author{
			FirstName: firstName,
			LastName:  lastName,
		},
	})
}
